process/track/blockProcessor.go

```diff
@@ -233,31 +241,83 @@ func (bp *blockProcessor) requestHeadersIfNeeded(
 		return
 	}
 
-	nbSortedHeaders := len(sortedHeaders)
-	if nbSortedHeaders == 0 {
+	shouldRequestHeaders := false
+
+	defer func() {
+		if !shouldRequestHeaders {
+			bp.requestHeadersIfNothingNewIsReceived(lastNotarizedHeader, sortedHeaders, longestChainHeaders)
+		}
+	}()
+
+	numSortedHeaders := len(sortedHeaders)
+	if numSortedHeaders == 0 {
 		return
 	}
 
-	highestNonceReceived := sortedHeaders[nbSortedHeaders-1].GetNonce()
+	highestNonceReceived := sortedHeaders[numSortedHeaders-1].GetNonce()
 	highestNonceInLongestChain := lastNotarizedHeader.GetNonce()
-	nbLongestChainHeaders := len(longestChainHeaders)
-	if nbLongestChainHeaders > 0 {
-		highestNonceInLongestChain = longestChainHeaders[nbLongestChainHeaders-1].GetNonce()
+	numLongestChainHeaders := len(longestChainHeaders)
+	if numLongestChainHeaders > 0 {
+		highestNonceInLongestChain = longestChainHeaders[numLongestChainHeaders-1].GetNonce()
 	}
 
-	shouldRequestHeaders := highestNonceReceived > highestNonceInLongestChain+bp.blockFinality && nbLongestChainHeaders == 0
+	shouldRequestHeaders = highestNonceReceived > highestNonceInLongestChain+bp.blockFinality && numLongestChainHeaders == 0
```

Comment:
why aren't we requesting if we have a not empty longestChainHeaders?

Reply:
This is done to avoid many requests when node is in sync mode and it receives headers with higher nonces than it's last processed header nonce. As long as it can advance, so numLongestChainHeaders > 0 than, it should not request for headers in advance